src/librustc_mir/transform/match_branches.rs

```diff
@@ -42,50 +69,67 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
-                        if let Some(f_c) = f_c.literal.try_eval_bool(tcx, param_env) {
-                            // This should also be a bool because it's writing to the same place
-                            let s_c = s_c.literal.try_eval_bool(tcx, param_env).unwrap();
-                            if f_c != s_c {
-                                // have to check this here because f_c & s_c might have
-                                // different spans.
-                                continue;
-                            }
-                        }
-                        continue 'outer;
-                    }
-                    // If there are not exclusively assignments, then ignore this
+                    ) if lhs_f == lhs_s
+                        && f_c.literal.ty.is_bool()
+                        && s_c.literal.ty.is_bool()
+                        && f_c.literal.try_eval_bool(tcx, param_env).is_some()
+                        && s_c.literal.try_eval_bool(tcx, param_env).is_some() => {}
+
+                    // Otherwise we cannot optimize. Try another block.
                     _ => continue 'outer,
                 }
             }
-            // Take owenership of items now that we know we can optimize.
+            // Take ownership of items now that we know we can optimize.
             let discr = discr.clone();
-            let (from, first) = bbs.pick2_mut(bb_idx, first);
 
-            let new_stmts = first.statements.iter().cloned().map(|mut s| {
-                if let StatementKind::Assign(box (_, ref mut rhs)) = s.kind {
-                    if let Rvalue::Use(Operand::Constant(c)) = rhs {
-                        let size = tcx.layout_of(param_env.and(switch_ty)).unwrap().size;
-                        let const_cmp = Operand::const_from_scalar(
-                            tcx,
-                            switch_ty,
-                            crate::interpret::Scalar::from_uint(val, size),
-                            rustc_span::DUMMY_SP,
-                        );
-                        if let Some(c) = c.literal.try_eval_bool(tcx, param_env) {
-                            let op = if c { BinOp::Eq } else { BinOp::Ne };
-                            *rhs = Rvalue::BinaryOp(op, Operand::Move(discr), const_cmp);
+            let new_stmts = first_stmts
+                .iter()
+                .zip(scnd_stmts.iter())
+                .map(|(f, s)| {
+                    match (&f.kind, &s.kind) {
+                        (f_s, s_s) if f_s == s_s => (*f).clone(),
+
+                        (
+                            StatementKind::Assign(box (lhs, Rvalue::Use(Operand::Constant(f_c)))),
+                            StatementKind::Assign(box (_, Rvalue::Use(Operand::Constant(s_c)))),
+                        ) => {
+                            let f_b = f_c.literal.try_eval_bool(tcx, param_env).unwrap();
+                            let s_b = s_c.literal.try_eval_bool(tcx, param_env).unwrap();
```

Comment:
Will this panic if there are other assignment statements which are not bools?
i.e. `a = 3, b = 4`

Reply:
ah no nvm, this shouldn't panic because it would've matched earlier and skipped this block

Reply:
I added a comment explaining why unwrap is fine.